packages/@sfx/sayt-driver-plugin/src/sayt-driver-plugin.ts

```diff
@@ -120,6 +174,55 @@ export default class SaytDriverPlugin implements Plugin {
     return [searchTerms];
   }
 
+  /**
+   * Extracts query and products from the given response.
+   *
+   * @param response An object containing the original query and product records.
+   * @returns An object containing the query and an array of valid simplified products.
+   */
+  searchCallback(response: Results): ProductsResponseSection {
+    const { query, records } = response;
+    const mappedRecords = records.map(record => {
+      const filter = this.parseRecord(record);
+      if (filter === undefined) return;
+      const { data, firstVariant, nonvisualVariants } = filter;
+      return {
+        title: data.title,
+        price: nonvisualVariants[0].originalPrice,
+        imageSrc: firstVariant.productImage,
+        imageAlt: data.title,
+        productUrl: firstVariant.productImage,
+        // @TODO Handle variants
+      }
+    }).filter(record => record);
+    return {
+      query,
+      products: mappedRecords,
+    };
+  }
+
+  /**
+   * Parses a given product record for valid data.
+   *
+   * @param record An object containing the product record data.
+   * @returns An object containing relevant product data,
+   * or undefined if record is invalid.
+   */
+  parseRecord(record: Record): any | undefined {
```

Comment:
`any | undefined` is redundant. Since you know the shape of the data you're returning, `any` should be a concrete interface.

Furthermore, since it looks like `undefined` is being used as an error case, I think this is a prime case for throwing an exception.

With an exception, the function body simplifies to:

```js
const data = record.allMeta;
const firstVariant = data.visualVariants[0];
const nonvisualVariants = firstVariant.nonvisualVariants;

if (!nonVisualVariants[0]) throw new Error('No nonvisual variants');

return {
  data,
  firstVariant,
  nonvisualVariant,
};
```

Reply:
I like the exception idea. I do not really want to use an interface for this, however. Frankly, this code is so likely to change that adding an interface seems like make-work. It's a weird function to begin with. There will be a story to develop the transformation of products. At that time, we can update the relevant functions from this PR and create some meaningful interfaces.

Reply:
In that case, just `any` is sufficient.